src/OrchardCore/OrchardCore.DisplayManagement/Implementation/DefaultHtmlDisplay.cs

```diff
@@ -237,44 +239,50 @@ private async Task<ShapeBinding> GetShapeBindingAsync(string shapeType, Alternat
             }
 
             // When no alternates matches, the shapeType is used to find the longest matching binding,
-            // the shape-type name can break itself into shorter fallbacks at double-underscore marks,
-            // so the shape-type itself may contain a longer alternate forms that falls back to a shorter one.
-            var shapeTypeScan = shapeType;
+            // the shapetype name can break itself into shorter fallbacks at double-underscore marks,
+            // so the shapetype itself may contain a longer alternate forms that falls back to a shorter one.
 
-            do
+            // Build a cache of such values
+            var alternateShapeTypes = _alternateShapeTypes.GetOrAdd(shapeType, shapeType =>
```

Comment:
Did you benchmark this change?

Reply:
With bombardier locally by removing other components, and it showed. Also I did it as it appeared as a main factor of string concatenation, then looked into the TagHelper implementation which is creating new string of every call.